engines/twine/menu.cpp

```diff
@@ -0,0 +1,1214 @@
+			if (!_engine->_keyboard.pressedKey && !_engine->_keyboard.skippedKey) {
+				di = 0;
+			}
+		}
+
+		if (_engine->loopCurrentKey == 1 || _engine->loopPressedKey & 0x20)
+			break;
+
+		if (_engine->_keyboard.key & 2) { // down
+			inventorySelectedItem++;
+			if (inventorySelectedItem >= NUM_INVENTORY_ITEMS) {
+				inventorySelectedItem = 0;
+			}
+			drawItem(prevSelectedItem);
+			bx = 3;
+		}
+
+		if (_engine->_keyboard.key & 1) { // up
+			inventorySelectedItem--;
+			if (inventorySelectedItem < 0) {
+				inventorySelectedItem = NUM_INVENTORY_ITEMS - 1;
+			}
+			drawItem(prevSelectedItem);
+			bx = 3;
+		}
+
+		if (_engine->_keyboard.key & 4) { // left
+			inventorySelectedItem -= 4;
+			if (inventorySelectedItem < 0) {
+				inventorySelectedItem += NUM_INVENTORY_ITEMS;
+			}
+			drawItem(prevSelectedItem);
+			bx = 3;
+		}
+
+		if (_engine->_keyboard.key & 8) { // right
+			inventorySelectedItem += 4;
+			if (inventorySelectedItem >= NUM_INVENTORY_ITEMS) {
+				inventorySelectedItem -= NUM_INVENTORY_ITEMS;
+			}
+			drawItem(prevSelectedItem);
+			bx = 3;
+		}
+
+		if (bx == 3) {
+			_engine->_text->initInventoryDialogueBox();
+
+			if (_engine->_gameState->gameFlags[inventorySelectedItem] == 1 && !_engine->_gameState->gameFlags[GAMEFLAG_INVENTORY_DISABLED] && inventorySelectedItem < NUM_INVENTORY_ITEMS) {
+				_engine->_text->initText(inventorySelectedItem + 100);
+			} else {
+				_engine->_text->initText(128);
+			}
+			bx = 0;
+		}
+
+		if (bx != 2) {
+			bx = _engine->_text->printText10();
+		}
+
+		// TRICKY: 3D model rotation delay - only apply when no text is drawing
+		if (bx == 0 || bx == 2) {
+			_engine->_system->delayMillis(15);
+		}
+
+		if (_engine->loopPressedKey & 1) {
+			if (bx == 2) {
+				_engine->_text->initInventoryDialogueBox();
+				bx = 0;
+			} else {
+				if (_engine->_gameState->gameFlags[inventorySelectedItem] == 1 && !_engine->_gameState->gameFlags[GAMEFLAG_INVENTORY_DISABLED] && inventorySelectedItem < NUM_INVENTORY_ITEMS) {
+					_engine->_text->initInventoryDialogueBox();
+					_engine->_text->initText(inventorySelectedItem + 100);
+				}
+			}
+		}
+
+		drawItem(inventorySelectedItem);
+
+		if ((_engine->loopPressedKey & 2) && _engine->_gameState->gameFlags[inventorySelectedItem] == 1 && !_engine->_gameState->gameFlags[GAMEFLAG_INVENTORY_DISABLED] && inventorySelectedItem < NUM_INVENTORY_ITEMS) {
+			_engine->loopInventoryItem = inventorySelectedItem;
+			inventorySelectedColor = 91;
+			drawItem(inventorySelectedItem);
+			break;
+		}
+	}
+
+	_engine->_text->printTextVar13 = 0;
+
+	_engine->_scene->alphaLight = tmpAlphaLight;
+	_engine->_scene->betaLight = tmpBetaLight;
+
+	_engine->_gameState->initEngineProjections();
+
+	_engine->cfgfile.LanguageCDId = tmpLanguageCD;
+
+	_engine->_text->initTextBank(_engine->_text->currentTextBank + 3);
+
+	while (_engine->_keyboard.skipIntro != 0 && _engine->_keyboard.skippedKey != 0) {
+		readKeys();
+		_engine->_system->delayMillis(1);
```

Comment:
you potentially need also to call `g_system->updateScreen()`, so mouse gets redrawn.